### PR TITLE
Feat save only requested detector and pixels to fits

### PR DESCRIPTION
**Pull request by samaloney**
When investigating #291 it was clear for some specific requests the files are 31 time larger than they need to be (when only the BKG detector) is requested. Further as many of our pixel data request are for large pixels only the are 1/3 larger than necessary. 

This PR subscrits the count array in the detector, pixel and engery dimensions so the files are as small as possible.

**This change would require updates to the python and IDL analysis software**

### CODE REVIEW

File: stixcore/products/level0/scienceL0.py [hunk 1]
```diff
@@ -445,41 +435,21 @@ def from_levelb(cls, levelb, parent=''):
...
-            raise ValueError('Original and reformatted count totals do not match')
+            if counts.sum() != rebinned_counts.sum():
+                raise ValueError('Original and reformatted count totals do not match')
+            counts = rebinned_counts
+            counts_var = rebinned_counts_var
+
+        # check pixel mask and subscript down to match max
+        cids, *_ = np.where(data['pixel_masks'].sum(axis=0) > 0)
+        counts = counts[..., cids, :]
+        counts_var = counts_var[..., cids, :]
```
Comment (nicHoch) [inline stixcore/products/level0/scienceL0.py:446]: this is nice compact code

**Review by nicHoch — APPROVED** [review]
good to go

also e2e looks like that only dimensions pixels/detectors are touched

**Comment by samaloney** [comment]
The e2e test look odd to me to many changes so looked into it an there were  a few issue but think I have them all sorted now

**Comment by FredSchuller** [comment]
I've run a similar test as described in #291 on the new test files recently generated by @nicHoch (I only looked at the cpd files). There are two for which the number of energy bins is not as expected:
`solo_L1_stix-sci-xray-cpd_20220912T070823-20220912T071528_V01_2209126204-50199.fits --  N_energy  8 / 32>`
`solo_L1_stix-sci-xray-cpd_20220917T130244-20220917T131255_V01_2209176255-50200.fits --  N_energy  8 / 32>`
(the energy_bin_mask in the CONTROL table looks weird... the counts shape pretends that we were using all 32 energy bins, but this doesn't seem to be the case)

However, no file for which I detected a mismatch in the number of detectors last year is part of this re-processing.
Could you for example re-process these three files, for which the previous version was showing the issue:
`solo_L1_stix-sci-xray-cpd_20210310T202144-20210310T202515_V01_2103100003-53351.fits --  N_det 30 / 32`

`solo_L1_stix-sci-xray-cpd_20210419T232707-20210419T232739_V01_2104190009-53705.fits --  N_pix  4 / 12 N_det  4 / 32`

`solo_L1_stix-sci-xray-cpd_20210713T174459-20210713T210459_V01_2107130043-51017.fits --  N_det  1 / 32`

**Comment by nicHoch** [comment]
@FredSchuller I already reprocessed "everything". The entire data set is now available here:

http://pub099.cs.technik.fhnw.ch/data/testing/esa_release/

**Comment by FredSchuller** [comment]
@nicHoch OK, then I will download everything and have a look.
Could you maybe delete the folders 2068/, 2088/ and 2090/ ? I guess you don't want to release these to ESA...

**Comment by FredSchuller** [comment]
So, it seems that only some issues have been fixed, but also new issues have appeared...
Attached is a list of file names with mismatch in at least one dimension.
Examples of new issues:

- when only one detector is used (e.g. BKG), the shape of counts array has one dimension less, at least in some cases. But I'm not sure if it's a problem only in the IDL reading procedure, since the TDIM11 keyword in DATA table header looks ok (for example (22,12,1) in file `solo_L1_stix-sci-xray-cpd_20220519T115032-20220519T122347_V01_2205197852-52631.fits`
- the pixel mask changes within a file, so that the shape of counts array is as expected for some part, but not all (e.g. file `solo_L1_stix-sci-xray-cpd_20220906T100000-20220906T150001_V01_2209066375-61484.fits`
- energy mask is not consistent with a contiguous range of energy bins, e.g. files:
`solo_L1_stix-sci-xray-cpd_20220912T070823-20220912T071528_V01_2209126204-50199.fits`
`solo_L1_stix-sci-xray-cpd_20220917T130244-20220917T131255_V01_2209176255-50200.fits`

[stx_cpd_files_esa_release_bad_shape.txt](https://github.com/i4Ds/STIXCore/files/10687448/stx_cpd_files_esa_release_bad_shape.txt)

**Comment by samaloney** [comment]
Thanks for looking @FredSchuller  may I ask how you are generating this list/comparison? I just quickly looked:
* `solo_L1_stix-sci-xray-cpd_20220519T115032-20220519T122347_V01_2205197852-52631.fits`
    ```
    l1 = fits.open('http://pub099.cs.technik.fhnw.ch/data/testing/esa_release/L1/2022/05/19/SCI/solo_L1_stix-sci-xray-cpd_20220519T115032-20220519T122347_V01_2205197852-52631.fits')
    l1[2].data['counts'].shape
    (314, 1, 12, 22)
    l1[2].data['detector_masks'][0]
    array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=uint8)
    ```
 * `solo_L1_stix-sci-xray-cpd_20220906T100000-20220906T150001_V01_2209066375-61484.fits` and the pixel_mask and counts are behaving as expected. In the case where the pixel_mask changes during a request we have to use the largest pixel_mask and the position of the counts matches the mask
   ```
   l1 = fits.open('http://pub099.cs.technik.fhnw.ch/data/testing/esa_release/L1/2022/09/06/SCI/solo_L1_stix-sci-xray-cpd_20220906T100000-20220906T150001_V01_2209066375-61484.fits')

   l1[2].data['counts'].sum(axis=(1, 3))[0]
   array([2545943, 2468478, 2481511, 2522134, 2515100, 2464646, 2449159,
       2540662,       0,       0,       0,       0], dtype=uint64)

   l1[2].data['pixel_masks'][0]
   array([1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0], dtype=uint8

   l1[2].data['counts'].sum(axis=(1, 3))[-1]
   array([419224, 401563, 401692, 410520, 410186, 401264, 399452, 411538,
        30905,  28301,  30106,  28493], dtype=uint64)
   
   l1[2].data['pixel_masks'][-1]
   array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=uint8)

   np.all(l1[2].data['pixel_masks'].astype(bool) == (l1[2].data['counts'].sum(axis=(1, 3)) > 0))
   True
   ```

* the last one is related to the use of onboard summing which I didn't know we were actively using so hadn't really considered it. Will need to discuss on Thursday as requires a larger change to the fits files.

Also I think it would be useful to move this discussion to the relevant issue or a new issue not this closed PR

**Pull request merged**